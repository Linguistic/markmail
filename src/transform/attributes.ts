import { mapKeys } from "@std/collections/map_keys.ts";
import { paramCase } from "case";

import { ConfigSchema, Typography } from "../config/mod.ts";
import { compact, range } from "./utils.ts";

export function createHeaderClass(depth: number, config: Typography) {
  const {
    color,
    font,
    leading,
    tracking,
    height,
    textDecoration,
    textTransform,
    align,
    containerBackgroundColor,
    padding,
  } = config;

  const properties = compact({
    ...mapKeys(font.size, paramCase),
    ...mapKeys(padding, (v) => `padding-${v}`),
    color,
    align,
    height,
    "font-family": font.family,
    "font-style": font.style,
    "font-weight": font.weight,
    "line-height": leading,
    "letter-spacing": tracking,
    "text-decoration": textDecoration,
    "text-transform": textTransform,
    "container-background-color": containerBackgroundColor,
  });

  const className = `h${depth}`;

  const propertyString = Object.entries(properties).map(([key, value]) =>
    `${key}="${value}"`
  ).join(" ");

  return `<mj-class name="${className}" ${propertyString} />`;
}

export function createAttributes(config: ConfigSchema) {
  // Header classes
  const headerClasses = range(1, 6).map((i) =>
    createHeaderClass(
      i,
      config.theme.headings[`h${i}` as keyof typeof config.theme.headings],
    )
  );

  return [headerClasses].join("\n");
}
