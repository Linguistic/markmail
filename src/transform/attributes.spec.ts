import { describe, it } from "@std/testing/bdd.ts";
import { assertSnapshot } from "@std/testing/snapshot.ts";

import { createHeaderClass } from "./attributes.ts";

describe("createHeaderClass", () => {
  it("creates a header class correctly", async (t) => {
    await assertSnapshot(
      t,
      createHeaderClass(1, {
        color: "red",
        font: {
          family: "sans-serif, monospace",
          size: {
            fontSize: "1rem",
            lineHeight: "2rem",
          },
          style: "italic",
          weight: 600,
        },
        leading: "2rem",
        tracking: 12,
        height: 1.5,
        textDecoration: "underline",
        textTransform: "uppercase",
        align: "left",
        containerBackgroundColor: "blue",
        padding: {
          top: "1rem",
          bottom: "1rem",
          left: "1rem",
          right: "1rem",
        },
      }),
    );
  });
});
