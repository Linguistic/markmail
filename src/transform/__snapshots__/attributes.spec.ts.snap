export const snapshot = {};

snapshot[`createHeaderClass > creates a header class correctly 1`] = `'<mj-class name="h1" font-size="1rem" line-height="2rem" padding-top="1rem" padding-bottom="1rem" padding-left="1rem" padding-right="1rem" color="red" align="left" height="1.5" font-family="sans-serif, monospace" font-style="italic" font-weight="600" letter-spacing="12" text-decoration="underline" text-transform="uppercase" container-background-color="blue" />'`;
